moirae/extractors/ecm.py

```diff
@@ -256,3 +257,277 @@ def extract(self, dataset: CellDataset) -> Resistance:
+
+            # find the state (charging or discharging) of the previous step
+            if 'charging' in step_data['state'].values:
+                sel = step_data['state'] == 'charging'
+                state = 'ch'
+            elif 'discharging' in step_data['state'].values:
+                sel = step_data['state'] == 'discharging'
+                state = 'di'
+            else:
+                step_data_prev = step_data.copy()
+                print('no charging or discharing in previous step\n')
+                continue
+
+            step_data_chdi = step_data.loc[sel]
+
+            if len(step_data_chdi) > 0:
+                indx_end_chdi = step_data_chdi.index[-1]
+            else:
+                indx_end_chdi = step_data.index[0]
+
+            tmp = (step_data['current'].loc[indx_end_chdi + 1:].abs()
+                   > 1e-5).cumsum() == 0
+            indx_rest = tmp[tmp].index
+            t_rest = step_data['test_time'][indx_rest]
+
+            t_rest_el = t_rest.max()-t_rest.min()
+            if t_rest_el < self.min_rest:
+                step_data_prev = step_data.copy()
+                continue
+
+            socs.append(soc)
+
+            rests.append({
+                'soc': soc,
+                'step_data': step_data,
+                'indx_rest': indx_rest,
+                't_rest': t_rest,
+                'state': state,
+                'Iprev': Iprev
+                })
+
+        rests_ = [rests[ii] for ii in np.argsort(socs)]
+
+        return rests_
+
+    def _error(self, params, cycle_data, indx_fitseg, t_fitseg, state):
+        """Calculate the prediction MSE for the RC elements of a cell
+
+        Args:
+            params: The parameters of the RC pairs
+            cycle_data: Dataset containing the time series measurement
+            indx_fitseg: Time index for the fitting segment (the rest)
+            t_fitset: Time sequence for the fitting segment
+            state: State of the previous segment (ch or di)
+        Returns:
+            An tuple of RC instances with the requested SOC interpolation points
+        """
+        npset = params.shape[0]
+
+        As = params[0::2]
+        Ts = params[1::2]
+        for i_rc in np.arange(self.n_rc-1)[::-1]:
+            Ts[i_rc] *= Ts[i_rc+1]
+
+        t = np.array(t_fitseg) - np.array(t_fitseg)[0]
+        t = t[..., None]
+
+        res = cycle_data['voltage'][indx_fitseg[-1]]
+        res *= np.ones((t.size, npset))
+
+        if state == 'di':
+            for i_rc in range(self.n_rc):
+                res -= As[i_rc]*np.exp(-t/Ts[i_rc])
+
+        elif state == 'ch':
+            for i_rc in range(self.n_rc):
+                res += As[i_rc]*np.exp(-t/Ts[i_rc])
+
+        ref = cycle_data['voltage'][indx_fitseg]
+        ref = np.array(ref)[:, None]
+
+        err = np.sum((ref-res)**2, axis=0)
+
+        err[np.isnan(err)] = np.inf
+
+        return np.sum(err)
+
+    def extract(self, dataset: CellDataset) -> Resistance:

```

Comment:
Could you fix the return type for this argument? It's a tuple of RCComponents, right?

Reply:
done!